pkg/aggregator/ckey/key.go

```diff
@@ -27,49 +23,93 @@ import (
+// generating the hash.
+// Not safe for concurrent usage.
+type KeyGenerator struct {
+	// reused buffer to not create a uint64 on the stack every key generation
+	intb uint64
+
+	// seen is used as a hashset to deduplicate the tags when there is more than
+	// 16 and less than 512 tags.
+	seen []uint64
+	// empty is an empty hashset with all values set to 0, to reset `seen`
+	empty []uint64
+
+	// idx is used to deduplicate tags when there is less than 16 tags (faster than the
+	// hashset) or more than 512 tags (hashset has been allocated with 512 values max)
+	idx int
+}
+
 // Generate returns the ContextKey hash for the given parameters.
 // The tags array is sorted in place to avoid heap allocations.
 func (g *KeyGenerator) Generate(name, hostname string, tags []string) ContextKey {
-	g.buf = g.buf[:0]
+	g.intb = 0xc6a4a7935bd1e995
+	g.intb = g.intb ^ murmur3.StringSum64(name)
+	g.intb = g.intb ^ murmur3.StringSum64(hostname)
 
-	// Sort the tags in place. For typical tag slices, we use
-	// the in-place insertion sort to avoid heap allocations.
-	// We default to stdlib's sort package for longer slices.
-	// See `pkg/util/sort.go` for info on the threshold.
-	if len(tags) < util.InsertionSortThreshold {
-		util.InsertionSort(tags)
+	// there is two implementations used here to deduplicate the tags depending on how
+	// many tags we have to process:
+	//   -  16 < n < 512: 		we use a hashset of 512 values. This size has been
+	// 							selected to have space for approximately 500 tags
+	// 							since it nos impacting much the performances,
+	// 							even if the backend is truncating after 100.
+	//   -  n < 16 or n > 512: 	we use a simple for loops, which is faster than
+	//                         	the hashset when there is less than 16 tags, and
+	//                         	we use it as fallback when there is more than 512
+	//                         	because it is the maximum size the allocated
+	//                         	hashset can handle.
+	if len(tags) > 16 && len(tags) < 512 {
+		copy(g.seen, g.empty) // reset the `seen` hashset
+	OUTER:
+		for i := range tags {
+			h := murmur3.StringSum64(tags[i])
+			j := h & 511 // address this hash into the hashset
+			for {
+				if g.seen[j] == 0 { // not seen, we will add it to the hash
+					g.seen[j] = h
+					break
+				} else if g.seen[j] == h {
+					continue OUTER // already seen, we do not want to xor multiple times the same tag
+				} else {
+					// move 'right' in the hashset because there is already a value,
+					// in this bucket, which is not the one we're dealing with right now,
+					// we may have already seen this tag
+					j = (j + 1) & 511
+				}
+			}
+			g.intb = g.intb ^ h // add this tag into the hash
+		}
 	} else {
-		sort.Strings(tags)
-	}
-
-	g.buf = append(g.buf, name...)
-	g.buf = append(g.buf, ',')
-	for i := 0; i < len(tags); i++ {
-		g.buf = append(g.buf, tags[i]...)
-		g.buf = append(g.buf, ',')
+		g.idx = 0
```

Comment:
Let me benchmark it, it's a member for historical purpose (was inspired by `g.buf`) but technically it can live on the stack, I'll bench this and update accordingly.

Reply:
Micro-benchs show that this version (member of the struct) is performing slightly better:

```
KeyGeneration/1-tags-8    21.0ns ± 1%  21.9ns ± 2%   +4.41%  (p=0.008 n=5+5)
KeyGeneration/2-tags-8    28.6ns ± 1%  31.1ns ± 5%   +8.61%  (p=0.008 n=5+5)
KeyGeneration/4-tags-8    43.8ns ± 1%  47.1ns ± 5%   +7.54%  (p=0.008 n=5+5)
KeyGeneration/8-tags-8    83.3ns ± 1%  88.8ns ± 3%   +6.59%  (p=0.008 n=5+5)
KeyGeneration/16-tags-8    193ns ± 4%   213ns ± 4%  +10.36%  (p=0.008 n=5+5)
```